src/rules.rs

```diff
@@ -146,193 +147,105 @@ mod tests {
         return rules;
     }
 
-    #[test]
-    // Test the type rule for a subject that is in the rules & type map
-    fn is_subject_set() {
-        let subject = Subject::NamedNode(NamedNode {
-            iri: "http://example.org/Alice".to_string(),
-        });
-        let rules = set_type_rule("Person");
-        let mut type_map = HashMap::new();
-        type_map.insert("http://example.org/Alice".to_string(), "Person".to_string());
-        let mut mask = TripleMask::default();
-        mask = match_type_rule_subject(&subject, mask, &type_map, &rules);
-        assert!(mask.is_set(&TripleMask::SUBJECT));
-        assert!(!mask.is_set(&TripleMask::OBJECT));
-    }
-    #[test]
-    // Test the type rule for a subject that is not in the rules but in the type map
-    fn subject_not_in_rules() {
-        let subject = Subject::NamedNode(NamedNode {
-            iri: "http://example.org/Alice".to_string(),
-        });
-        let rules = set_type_rule("Bank");
-        let mut type_map = HashMap::new();
-        type_map.insert("http://example.org/Alice".to_string(), "Person".to_string());
-        let mut mask = TripleMask::default();
-
-        mask = match_type_rule_subject(&subject, mask, &type_map, &rules);
-
-        assert!(!mask.is_set(&TripleMask::SUBJECT));
-        assert!(!mask.is_set(&TripleMask::OBJECT));
-    }
-
-    #[test]
-    // Test the type rule for a subject neither in the rules nor in the type map
-    fn subject_not_in_types() {
-        let subject = Subject::NamedNode(NamedNode {
-            iri: "http://example.org/Alice".to_string(),
-        });
-        let rules = set_type_rule("Bank");
-        let mut type_map = HashMap::new();
-        type_map.insert("http://example.org/Bank".to_string(), "Bank".to_string());
-        let mut mask = TripleMask::default();
-
-        mask = match_type_rule_subject(&subject, mask, &type_map, &rules);
-
-        assert!(!mask.is_set(&TripleMask::SUBJECT));
-        assert!(!mask.is_set(&TripleMask::OBJECT));
-    }
-    #[test]
-    // Test the type rule for an object that is in the rules & type map
-    fn is_object_set() {
-        let object = Term::NamedNode(NamedNode {
-            iri: "http://example.org/Alice".to_string(),
-        });
-        let rules = set_type_rule("Person");
-        let mut type_map = HashMap::new();
-        type_map.insert("http://example.org/Alice".to_string(), "Person".to_string());
+    #[rstest]
+    // Subject is in the rules & type index
+    #[case(true, "Alice", "Alice", "Person", "Person", true, false)]
+    // Subject is in the type index, not in the rules
+    #[case(true, "Alice", "Alice", "Person", "Bank", false, false)]
+    // Subject is not in the type index
+    #[case(true, "Alice", "BankName", "Bank", "Bank", false, false)]
+    // Object is in the rules & type index
+    #[case(false, "Alice", "Alice", "Person", "Person", false, true)]
+    fn type_rule(
```

Comment:
**suggestion:** same as  below, if `expected_o` is never be affected by this rule, we could drop arguments.